### DISCUSSION TITLE
Use F# for it's ability to compile to multiple languages via Fable (and great .NET language for scripting)

### DISCUSSION THREAD

**Discussion by jkone27** [microsoft/semantic-kernel#5419] [Ideas]
Just as a suggestion/idea, but since semantic kernel needs many client libraries for different languages, maybe a smart way to achieve this is to use a .NET language which actually can compile a big portion of its subset to other languages?

ofc the core could be kept ".NET" and also tests run in .NET maybe but different client/languages/variants could be compiled/transpiled to target languages?

in addition to this F# is awesome and allows great REPL/interactive development with `.fsx` single file scripts and `#r "nuget:...` support, and execution via `dotnet fsi` included in `dotnet-sdk`

https://jkone27-3876.medium.com/compiling-f-to-python-a9839332298d

* python
* javascript
* typescript
* rust

for example, java is not covered but it would be already quite some saving and unification of APIs possibly?
* https://fable.io/docs/getting-started/your-first-fable-project.html
* https://jkone27-3876.medium.com/compile-f-to-rust-22cf5aa9021

**Comment by sophialagerkranspandey** [comment]
Thanks for sharing. Internally, we are looking at F#. We would like to learn from you and see if there are other SDK's you are using/seen used?

**Comment by jkone27** [comment]
Hello, in .NET the other sdk i can have in mind is ML.NET, btw

i think also part of github next team is lead by F# devs, and also the F# language creator probably you already know but just posting in case this can be useful for anyone

* https://githubnext.com/team/dsyme/ (creator of F#, principal github next researcher 🧑‍🔬 )
* creator of ionide vscode extension and F# integration for vscode: https://githubnext.com/team/krzysztof-cieslak/

also from ML.NET team,  is an F# developer with great experience as well: 
* https://dotnet.microsoft.com/en-us/apps/machinelearning-ai/ml-dotnet
* https://www.luisquintanilla.me/posts/case-fsharp-machine-learning

i think very interesting is the ability of F# to transpile to many different target languages via fable, while mantaining automatic generalized type inference and thus very lean code, plus having the pipeline operator great in data computations (like elixir or gleam recently)

* mantainer of fable: https://github.com/MangelMaxime
* fable python: https://github.com/dbrattli
* creator and old mantainer of fable: https://github.com/alfonsogarciacaro


some articles i wrote comparing F# to python:
* https://jkone27-3876.medium.com/f-5-like-a-pythonist-on-steroid-8f3038d6c9a
and one COMPILING F# to python (already linked above)
* https://jkone27-3876.medium.com/compiling-f-to-python-a9839332298d

F# and Python vid:
* philipp carter principal .NET/F# developer and F# hero 🦸 , author of source generators for C#: https://www.youtube.com/watch?v=_QnbV6CAWXc 
* F# content creator sir hamy: https://www.youtube.com/watch?v=b12MFKsOVC8

nice NDC videos:
* fable python: https://www.youtube.com/watch?v=deny3mBfccQ
* F# for C# devs: https://www.youtube.com/watch?v=tR4SyQOPvKs

**Comment by jkone27** [comment]
worth mentioning also the awesome type providers for OpenApi and Fsharp Data for e.g. json access
* https://fsprojects.github.io/FSharp.Data/
* https://mallibone.com/post/fsharp-scripting --> reference nuget packages in one .fsx file no need for fsproj/csproj
* https://blog.lucca.io/2022/05/19/fsharp-script (fsx scripting)
* swagger provider: https://fsprojects.github.io/SwaggerProvider/#/OpenApiClientProvider --> might be useful to conned to different BE for different models and magically infer types from live openapi specs of those models if provided

**Comment by jkone27** [comment]
maybe you want to check this one ? https://freymaurer.github.io/Siren/blog
 
I suspect this would make targeting multi language way easier for big libraries like semantic kernel ?

@sophialagerkranspandey 👆 @moonbox3 ? can also find Fable.Multiverse template on GitHub , do you want to check it in case? Why was this closed ? Cheers
